pmlc/conversion/gpu/gather_gpu_launch_funcs.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+
+#include "pmlc/conversion/gpu/pass_detail.h"
+
+using namespace mlir; // NOLINT[build/namespaces]
+
+namespace pmlc::conversion::gpu {
+namespace gpu = mlir::gpu;
+
+struct Move {
+  mlir::Operation *insertionOp;
+  std::vector<mlir::Operation *> ops;
+};
+
+class GatherGpuLaunchFuncsPass
+    : public GatherGpuLaunchFuncsPassBase<GatherGpuLaunchFuncsPass> {
+public:
+  void runOnOperation() override;
+
+private:
+  mlir::Operation *
+  gatherAllocOpsStartingFromOp(mlir::Operation *op,
+                               std::vector<mlir::Operation *> &ops);
+  void gatherAllocOpsBetweenConsecutiveLaunchOps(
+      mlir::Block &block, std::vector<std::shared_ptr<Move>> &moves);
+};
+
+mlir::Operation *GatherGpuLaunchFuncsPass::gatherAllocOpsStartingFromOp(
+    mlir::Operation *op, std::vector<mlir::Operation *> &ops) {
+  while (op != nullptr) {
+    if (mlir::isa<mlir::AllocOp>(op)) {
+      ops.push_back(op);
+    }
+    if (!mlir::isa<mlir::AllocOp>(op) && !mlir::isa<gpu::LaunchFuncOp>(op)) {
+      return op;
+    }
+    op = op->getNextNode();
+  }
+  return op;
+}
+
+void GatherGpuLaunchFuncsPass::gatherAllocOpsBetweenConsecutiveLaunchOps(
+    mlir::Block &block, std::vector<std::shared_ptr<Move>> &moves) {
+  mlir::Operation *op = &(*block.begin());
+  while (op != nullptr) {
```

Comment:
This still feels unnecessarily complex.  I may be oversimplifying but I read this code as doing:

```
op = advanceToNextLaunchOp()
while (op != nullptr)
  if LaunchOp:
    create a Move 
    move->insertionOp = op
    moves.push_back(move)
    op = op->getNextNode()
  else if AllocOp:
    moves.back()->ops.push_back(op)
    op = op->getNextNode()
  else
     op = advanceToNextLaunchOp()
```

Reply:
Revised:  I don't think you need the first `advanceToNextLaunchOp` call as it is handled in the else case.

Reply:
Thanks. Here we are not collecting from every LaunchOp, we are collecting from the first LaunchOp that are not following otherLaunchOps or AllocOp.
An example would be, change from 

```
AllocOp
LaunchOp
AllocOp
LaunchOp
AllocOp
LaunchOp
SomeOtherOp
SomeOtherOp
AllocOp
LaunchOp
AllocOp
LaunchOp
```
To
```
AllocOp
AllocOp
AllocOp
LaunchOp
LaunchOp
LaunchOp
SomeOtherOp
SomeOtherOp
AllocOp
AllocOp
LaunchOp
LaunchOp
```
Here we will generate two moves, the first move for the first three allocOps before the first LaunchOp...we have to skip the second and third launchOp, that is why I add function `gatherAllocOpsStartingFromOp`